# Online Retail Sales Analysis

## 1. Project Overview

### 1.1 Introduction

This project focuses on analyzing an online retail transaction dataset using Python. The purpose is to explore retail sales data, identify patterns, understand product performance, and extract useful business insights.

### 1.2 Objectives

* Explore and understand the retail transaction dataset.
* Investigate data quality and prepare the data for analysis.
* Calculate key sales performance metrics.
* Analyze product sales and revenue trends.
* Visualize findings and communicate actionable business insights.

### 1.3 Tools and Technologies

* **Python:** Primary programming language.
* **Pandas:** Data manipulation and analysis.
* **NumPy:** Numerical operations.
* **Matplotlib:** Data visualization.
* **Seaborn:** Statistical visualization.
* **Jupyter Notebook:** Interactive analysis environment.
* **Git and GitHub:** Version control and project publication.

### 1.4 Project Scope

The analysis covers transaction-level retail data, including products, quantities, unit prices, invoice dates, and countries. The project will focus on sales performance, data quality, product analysis, and time-based trends.

## 2. Dataset Description

### 2.1 Data Source

The dataset used in this project is the **Online Retail** dataset from the UCI Machine Learning Repository.

* **Dataset:** Online Retail
* **Source:** UCI Machine Learning Repository
* **Period:** December 2010 – December 2011
* **File format:** Excel (`.xlsx`)
* **Original size:** 541,909 transaction records
* **Number of attributes:** 8

### 2.2 Dataset Overview

The dataset contains transactional records from a UK-based online retail business. Each record represents a product line within an invoice.

### 2.3 Data Dictionary

| Column      | Description                                                                  |
| ----------- | ---------------------------------------------------------------------------- |
| InvoiceNo   | Unique invoice identifier; invoices beginning with C indicate cancellations. |
| StockCode   | Product identifier.                                                          |
| Description | Product name or description.                                                 |
| Quantity    | Number of units purchased or returned.                                       |
| InvoiceDate | Date and time of the transaction.                                            |
| UnitPrice   | Price per unit in GBP (£).                                                   |
| CustomerID  | Customer identifier, where available.                                        |
| Country     | Country associated with the transaction.                                     |

### 2.4 Analytical Scope

The dataset will be used to investigate sales revenue, product performance, monthly sales trends, geographic performance, and transaction data quality.

The dataset does not contain predefined product categories, profit margins, or geographic regions. These will not be assumed in the analysis.


## 3. Import Libraries and Load Data

### 3.1 Import Libraries

The required Python libraries are imported to support data manipulation, numerical calculations, and visualization.

### 3.2 Load the Dataset

The original Online Retail Excel file is loaded into a Pandas DataFrame. The original dataset is preserved so that any future cleaning operations can be performed on a separate copy.


In [2]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

# Display settings
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 20)
pd.set_option("display.float_format", "{:,.2f}".format)

print("Libraries imported successfully!")

Libraries imported successfully!


In [3]:
# Load the original dataset

df = pd.read_excel("../data/Online Retail.xlsx")

print("Dataset loaded successfully!")

Dataset loaded successfully!


## 4. Initial Data Inspection

### 4.1 Introduction

Initial data inspection helps understand the structure, size, and characteristics of the dataset before performing data cleaning and analysis.

The following checks will be performed:

* Display the first 10 records.
* Identify the number of rows and columns.
* Examine column names.
* Review the data types of each column.
* Generate basic descriptive statistics.

These checks provide an overview of the dataset and help identify areas that require further investigation.


In [4]:
# Display the first 10 records
df.head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,"17,850.00",United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,"17,850.00",United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,"17,850.00",United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,"17,850.00",United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,"17,850.00",United Kingdom
5,536365,22752,SET 7 BABUSHKA NESTING BOXES,2,2010-12-01 08:26:00,7.65,"17,850.00",United Kingdom
6,536365,21730,GLASS STAR FROSTED T-LIGHT HOLDER,6,2010-12-01 08:26:00,4.25,"17,850.00",United Kingdom
7,536366,22633,HAND WARMER UNION JACK,6,2010-12-01 08:28:00,1.85,"17,850.00",United Kingdom
8,536366,22632,HAND WARMER RED POLKA DOT,6,2010-12-01 08:28:00,1.85,"17,850.00",United Kingdom
9,536367,84879,ASSORTED COLOUR BIRD ORNAMENT,32,2010-12-01 08:34:00,1.69,"13,047.00",United Kingdom


In [5]:
# Number of rows and columns
print("Dataset dimensions:", df.shape)
print("Total rows:", df.shape[0])
print("Total columns:", df.shape[1])

Dataset dimensions: (541909, 8)
Total rows: 541909
Total columns: 8


In [6]:
# Display all column names
print("Column names:")
print(df.columns.tolist())

Column names:
['InvoiceNo', 'StockCode', 'Description', 'Quantity', 'InvoiceDate', 'UnitPrice', 'CustomerID', 'Country']


In [7]:
# Examine the data types
print(df.dtypes)

InvoiceNo              object
StockCode              object
Description            object
Quantity                int64
InvoiceDate    datetime64[us]
UnitPrice             float64
CustomerID            float64
Country                   str
dtype: object


In [8]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[us]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 33.1+ MB


In [9]:
# Summary statistics for numerical columns
display(df.describe())

,Quantity,InvoiceDate,UnitPrice,CustomerID
count,"541,909.00",541909,"541,909.00","406,829.00"
mean,9.55,2011-07-04 13:34:57.156386,4.61,"15,287.69"
min,"-80,995.00",2010-12-01 08:26:00,"-11,062.06","12,346.00"
25%,1.00,2011-03-28 11:34:00,1.25,"13,953.00"
50%,3.00,2011-07-19 17:17:00,2.08,"15,152.00"
75%,10.00,2011-10-19 11:27:00,4.13,"16,791.00"
max,"80,995.00",2011-12-09 12:50:00,"38,970.00","18,287.00"
std,218.08,NaN,96.76,"1,713.60"


### 4.2 Initial Observations

Based on the initial inspection:

1. The dataset contains 541,909 rows and 8 columns.
2. The dataset contains transaction identifiers, product information, quantities, prices, dates, customer IDs, and countries.
3. The `InvoiceDate` column is already stored in a datetime format.
4. The `Quantity` column is an integer, while `UnitPrice` is a floating-point number.
5. The dataset includes both numerical and categorical attributes.

These observations provide an initial understanding of the dataset. Missing values, duplicates, and unusual transactions will be investigated separately during the data-quality investigation stage.


## 5. Data Quality Investigation

### 5.1 Introduction

Data quality investigation is an important step in preparing data for reliable analysis. The purpose of this stage is to identify missing values, duplicate records, invalid dates, unusual quantities, incorrect prices, and cancelled transactions.

The investigation will help determine which records require cleaning and which unusual values represent legitimate business transactions.

The following areas will be examined:

* Missing values
* Duplicate records
* Unusual quantities
* Invalid or unusual prices
* Missing or invalid dates
* Cancelled transactions

The original dataset will remain unchanged during this investigation.


In [10]:
# Count missing values in each column
missing_count = df.isnull().sum()

# Calculate missing percentages
missing_percentage = (
    df.isnull().mean() * 100
).round(2)

# Create a summary table
missing_summary = pd.DataFrame({
    "Missing Count": missing_count,
    "Missing Percentage": missing_percentage
})

display(
    missing_summary.sort_values(
        "Missing Percentage",
        ascending=False
    )
)

,Missing Count,Missing Percentage
CustomerID,135080,24.93
Description,1454,0.27
StockCode,0,0.00
InvoiceNo,0,0.00
Quantity,0,0.00
InvoiceDate,0,0.00
UnitPrice,0,0.00
Country,0,0.00


In [11]:
# Examine records with missing customer IDs
missing_customers = df[df["CustomerID"].isna()]

print("Rows with missing CustomerID:",
      len(missing_customers))

display(missing_customers.head(10))

Rows with missing CustomerID: 135080


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.00,NaN,United Kingdom
1443,536544,21773,DECORATIVE ROSE BATHROOM BOTTLE,1,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1444,536544,21774,DECORATIVE CATS BATHROOM BOTTLE,2,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1445,536544,21786,POLKADOT RAIN HAT,4,2010-12-01 14:32:00,0.85,NaN,United Kingdom
1446,536544,21787,RAIN PONCHO RETROSPOT,2,2010-12-01 14:32:00,1.66,NaN,United Kingdom
1447,536544,21790,VINTAGE SNAP CARDS,9,2010-12-01 14:32:00,1.66,NaN,United Kingdom
1448,536544,21791,VINTAGE HEADS AND TAILS CARD GAME,2,2010-12-01 14:32:00,2.51,NaN,United Kingdom
1449,536544,21801,CHRISTMAS TREE DECORATION WITH BELL,10,2010-12-01 14:32:00,0.43,NaN,United Kingdom
1450,536544,21802,CHRISTMAS TREE HEART DECORATION,9,2010-12-01 14:32:00,0.43,NaN,United Kingdom
1451,536544,21803,CHRISTMAS TREE STAR DECORATION,11,2010-12-01 14:32:00,0.43,NaN,United Kingdom


In [12]:
missing_descriptions = df[
    df["Description"].isna()
]

print("Missing descriptions:",
      len(missing_descriptions))

display(missing_descriptions.head(10))

Missing descriptions: 1454


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.00,NaN,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.00,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.00,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.00,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.00,NaN,United Kingdom
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.00,NaN,United Kingdom
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.00,NaN,United Kingdom
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.00,NaN,United Kingdom
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.00,NaN,United Kingdom
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.00,NaN,United Kingdom


In [13]:
# Count exact duplicate rows
duplicate_count = df.duplicated().sum()

print("Exact duplicate rows:", duplicate_count)

Exact duplicate rows: 5268


In [14]:
duplicates = df[df.duplicated(keep=False)]

display(duplicates.head(20))

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
485,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,"17,908.00",United Kingdom
489,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,"17,908.00",United Kingdom
494,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,"17,908.00",United Kingdom
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,"17,908.00",United Kingdom
521,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,"17,908.00",United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,"17,908.00",United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,"17,908.00",United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,"17,908.00",United Kingdom
548,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,"17,920.00",United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,"17,920.00",United Kingdom


In [15]:
print(
    "Unique invoices:",
    df["InvoiceNo"].nunique()
)

print(
    "Rows with repeated invoice numbers:",
    df["InvoiceNo"].duplicated().sum()
)

Unique invoices: 25900
Rows with repeated invoice numbers: 516009


In [16]:
print("Minimum quantity:", df["Quantity"].min())
print("Maximum quantity:", df["Quantity"].max())

print("Negative quantities:", (df["Quantity"] < 0).sum())
print("Zero quantities:", (df["Quantity"] == 0).sum())

Minimum quantity: -80995
Maximum quantity: 80995
Negative quantities: 10624
Zero quantities: 0


In [17]:
display(
    df[df["Quantity"] < 0]
    [["InvoiceNo", "Description",
      "Quantity", "InvoiceDate", "Country"]]
    .head(20)
)

,InvoiceNo,Description,Quantity,InvoiceDate,Country
141,C536379,Discount,-1,2010-12-01 09:41:00,United Kingdom
154,C536383,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,United Kingdom
235,C536391,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,United Kingdom
236,C536391,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,United Kingdom
237,C536391,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,United Kingdom
238,C536391,PACK OF 12 RED RETROSPOT TISSUES,-24,2010-12-01 10:24:00,United Kingdom
239,C536391,CHICK GREY HOT WATER BOTTLE,-12,2010-12-01 10:24:00,United Kingdom
240,C536391,PLASTERS IN TIN VINTAGE PAISLEY,-12,2010-12-01 10:24:00,United Kingdom
241,C536391,PLASTERS IN TIN SKULLS,-24,2010-12-01 10:24:00,United Kingdom
939,C536506,JAM MAKING SET WITH JARS,-6,2010-12-01 12:38:00,United Kingdom


In [18]:
display(
    df.nlargest(10, "Quantity")[
        ["InvoiceNo", "Description",
         "Quantity", "UnitPrice"]
    ]
)

,InvoiceNo,Description,Quantity,UnitPrice
540421,581483,"PAPER CRAFT , LITTLE BIRDIE",80995,2.08
61619,541431,MEDIUM CERAMIC TOP STORAGE JAR,74215,1.04
502122,578841,ASSTD DESIGN 3D PAPER STICKERS,12540,0.00
74614,542504,NaN,5568,0.00
421632,573008,WORLD WAR 2 GLIDERS ASSTD DESIGNS,4800,0.21
206121,554868,SMALL POPCORN HOLDER,4300,0.72
220843,556231,?,4000,0.00
97432,544612,EMPIRE DESIGN ROSETTE,3906,0.82
270885,560599,ESSENTIAL BALM 3.5g TIN IN ENVELOPE,3186,0.06
52711,540815,FAIRY CAKE FLANNEL ASSORTED COLOUR,3114,2.10


In [19]:
print("Minimum price:", df["UnitPrice"].min())
print("Maximum price:", df["UnitPrice"].max())

print("Zero prices:", (df["UnitPrice"] == 0).sum())
print("Negative prices:", (df["UnitPrice"] < 0).sum())

Minimum price: -11062.06
Maximum price: 38970.0
Zero prices: 2515
Negative prices: 2


In [20]:
display(
    df[df["UnitPrice"] <= 0]
    [["InvoiceNo", "Description",
      "Quantity", "UnitPrice"]]
    .head(20)
)

,InvoiceNo,Description,Quantity,UnitPrice
622,536414,NaN,56,0.00
1970,536545,NaN,1,0.00
1971,536546,NaN,1,0.00
1972,536547,NaN,1,0.00
1987,536549,NaN,1,0.00
1988,536550,NaN,1,0.00
2024,536552,NaN,1,0.00
2025,536553,NaN,3,0.00
2026,536554,NaN,23,0.00
2406,536589,NaN,-10,0.00


In [21]:
display(
    df.nlargest(10, "UnitPrice")[
        ["InvoiceNo", "Description",
         "Quantity", "UnitPrice"]
    ]
)

,InvoiceNo,Description,Quantity,UnitPrice
222681,C556445,Manual,-1,"38,970.00"
524602,C580605,AMAZON FEE,-1,"17,836.46"
43702,C540117,AMAZON FEE,-1,"16,888.02"
43703,C540118,AMAZON FEE,-1,"16,453.71"
15016,C537630,AMAZON FEE,-1,"13,541.33"
15017,537632,AMAZON FEE,1,"13,541.33"
16356,C537651,AMAZON FEE,-1,"13,541.33"
16232,C537644,AMAZON FEE,-1,"13,474.79"
524601,C580604,AMAZON FEE,-1,"11,586.50"
299982,A563185,Adjust bad debt,1,"11,062.06"


In [22]:
print("Earliest invoice date:",
      df["InvoiceDate"].min())

print("Latest invoice date:",
      df["InvoiceDate"].max())

print("Missing invoice dates:",
      df["InvoiceDate"].isna().sum())

Earliest invoice date: 2010-12-01 08:26:00
Latest invoice date: 2011-12-09 12:50:00
Missing invoice dates: 0


In [23]:
print("InvoiceDate data type:",
      df["InvoiceDate"].dtype)

InvoiceDate data type: datetime64[us]


In [24]:
cancelled_mask = (
    df["InvoiceNo"]
    .astype(str)
    .str.startswith("C")
)

cancelled = df[cancelled_mask]

print("Cancelled invoice rows:", len(cancelled))

display(cancelled.head(10))

Cancelled invoice rows: 9288


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
141,C536379,D,Discount,-1,2010-12-01 09:41:00,27.50,"14,527.00",United Kingdom
154,C536383,35004C,SET OF 3 COLOURED FLYING DUCKS,-1,2010-12-01 09:49:00,4.65,"15,311.00",United Kingdom
235,C536391,22556,PLASTERS IN TIN CIRCUS PARADE,-12,2010-12-01 10:24:00,1.65,"17,548.00",United Kingdom
236,C536391,21984,PACK OF 12 PINK PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,"17,548.00",United Kingdom
237,C536391,21983,PACK OF 12 BLUE PAISLEY TISSUES,-24,2010-12-01 10:24:00,0.29,"17,548.00",United Kingdom
238,C536391,21980,PACK OF 12 RED RETROSPOT TISSUES,-24,2010-12-01 10:24:00,0.29,"17,548.00",United Kingdom
239,C536391,21484,CHICK GREY HOT WATER BOTTLE,-12,2010-12-01 10:24:00,3.45,"17,548.00",United Kingdom
240,C536391,22557,PLASTERS IN TIN VINTAGE PAISLEY,-12,2010-12-01 10:24:00,1.65,"17,548.00",United Kingdom
241,C536391,22553,PLASTERS IN TIN SKULLS,-24,2010-12-01 10:24:00,1.65,"17,548.00",United Kingdom
939,C536506,22960,JAM MAKING SET WITH JARS,-6,2010-12-01 12:38:00,4.25,"17,897.00",United Kingdom


In [25]:
cancellation_summary = pd.DataFrame({
    "Transaction Type": [
        "Cancelled",
        "Not Cancelled"
    ],
    "Rows": [
        cancelled_mask.sum(),
        (~cancelled_mask).sum()
    ]
})

display(cancellation_summary)

,Transaction Type,Rows
0,Cancelled,9288
1,Not Cancelled,532621


In [26]:
display(
    df.assign(
        IsCancelled=cancelled_mask,
        QuantityType=np.select(
            [
                df["Quantity"] < 0,
                df["Quantity"] == 0
            ],
            [
                "Negative",
                "Zero"
            ],
            default="Positive"
        )
    ).groupby(
        ["IsCancelled", "QuantityType"]
    ).size().unstack(fill_value=0)
)

QuantityType,Negative,Positive
IsCancelled,,
False,1336,531285
True,9288,0


In [27]:
quality_summary = pd.DataFrame({
    "Issue": [
        "Missing values (all columns)",
        "Exact duplicate rows",
        "Missing CustomerID",
        "Missing Description",
        "Missing InvoiceDate",
        "Negative quantities",
        "Zero quantities",
        "Zero prices",
        "Negative prices",
        "Cancelled invoice rows"
    ],
    "Count": [
        int(df.isnull().sum().sum()),
        int(df.duplicated().sum()),
        int(df["CustomerID"].isna().sum()),
        int(df["Description"].isna().sum()),
        int(df["InvoiceDate"].isna().sum()),
        int((df["Quantity"] < 0).sum()),
        int((df["Quantity"] == 0).sum()),
        int((df["UnitPrice"] == 0).sum()),
        int((df["UnitPrice"] < 0).sum()),
        int(cancelled_mask.sum())
    ]
})

display(quality_summary)

,Issue,Count
0,Missing values (all columns),136534
1,Exact duplicate rows,5268
2,Missing CustomerID,135080
3,Missing Description,1454
4,Missing InvoiceDate,0
5,Negative quantities,10624
6,Zero quantities,0
7,Zero prices,2515
8,Negative prices,2
9,Cancelled invoice rows,9288


### 5.2 Data Quality Findings

The investigation identified several data quality characteristics in the Online Retail dataset.

1. **Missing values:** The investigation identified missing values in CustomerID and Description. The most affected column was CustomerID with 135080 missing records.

2. **Duplicate records:** The dataset contained 5268 exact duplicate rows. Repeated invoice numbers were also observed, but these do not necessarily indicate duplicate transactions.

3. **Unusual quantities:** The dataset contained 10624 negative quantities and 0 zero quantities. Negative quantities require investigation because they may represent product returns.

4. **Unusual prices:** The investigation identified 2515 zero-priced records and 2 negative-priced records. These will be considered separately during cleaning.

5. **Dates:** The dataset covers the period from 2010-12-01 08:26:00 to 2011-12-09 12:50:00. There were 0 missing invoice dates.

6. **Cancellations:** The investigation identified 9288 cancelled invoice rows, which require separate treatment in sales analysis.

These findings will guide the cleaning process. Only records with a clear reason for removal will be excluded, while meaningful business transactions will be retained where possible.


## 6. Data Cleaning

### 6.1 Introduction

Data cleaning is the process of preparing raw data for reliable analysis. Based on the data-quality investigation, this stage will address duplicate records, missing values, unusual quantities, unusual prices, and cancelled transactions.

The cleaning process will preserve the original dataset and apply documented decisions to a separate working copy.

### 6.2 Cleaning Objectives

* Remove confirmed exact duplicate records.
* Handle records with missing essential information.
* Preserve legitimate returns and cancellations for investigation.
* Identify unusual prices and quantities without making unsupported assumptions.
* Validate the cleaned dataset.
* Save the cleaned data for subsequent analysis.


In [28]:
# Preserve the original dataset
df_clean = df.copy()

# Record the original number of rows
initial_rows = len(df_clean)

print("Original rows:", initial_rows)
print("Working copy created:", len(df_clean))

Original rows: 541909
Working copy created: 541909


In [29]:
# Count exact duplicates before cleaning
duplicates_before = df_clean.duplicated().sum()

# Remove exact duplicates
df_clean = df_clean.drop_duplicates().copy()

duplicates_removed = (
    initial_rows - len(df_clean)
)

print("Duplicates found:", duplicates_before)
print("Duplicates removed:", duplicates_removed)
print("Rows remaining:", len(df_clean))

Duplicates found: 5268
Duplicates removed: 5268
Rows remaining: 536641


In [30]:
# Record missing dates
missing_dates_before = df_clean["InvoiceDate"].isna().sum()

# Create a subset suitable for date-based analysis
df_dated = df_clean.dropna(
    subset=["InvoiceDate"]
).copy()

print("Rows with missing dates:", missing_dates_before)
print("Rows with valid dates:", len(df_dated))

Rows with missing dates: 0
Rows with valid dates: 536641


In [31]:
df_product = df_dated.dropna(
    subset=["Description"]
).copy()

print("Rows available for product analysis:",
      len(df_product))

Rows available for product analysis: 535187


In [32]:
# Identify unusual quantities and prices
df_clean["is_zero_quantity"] = (
    df_clean["Quantity"] == 0
)

df_clean["is_return"] = (
    df_clean["Quantity"] < 0
)

df_clean["is_zero_price"] = (
    df_clean["UnitPrice"] == 0
)

df_clean["is_negative_price"] = (
    df_clean["UnitPrice"] < 0
)

print(
    "Zero quantity:",
    df_clean["is_zero_quantity"].sum()
)

print(
    "Returns or negative quantities:",
    df_clean["is_return"].sum()
)

print(
    "Zero prices:",
    df_clean["is_zero_price"].sum()
)

print(
    "Negative prices:",
    df_clean["is_negative_price"].sum()
)

Zero quantity: 0
Returns or negative quantities: 10587
Zero prices: 2510
Negative prices: 2


In [33]:
# Identify cancelled invoices
df_clean["is_cancelled"] = (
    df_clean["InvoiceNo"]
    .astype(str)
    .str.startswith("C")
)

# Create an analysis-specific dataset
sales_clean = df_clean[
    (df_clean["Quantity"] > 0) &
    (df_clean["UnitPrice"] > 0) &
    (~df_clean["is_cancelled"]) &
    (df_clean["InvoiceDate"].notna()) &
    (df_clean["Description"].notna())
].copy()

print("Original rows:", initial_rows)
print("Cleaned sales rows:", len(sales_clean))

Original rows: 541909
Cleaned sales rows: 524878


In [34]:
# Check the cleaned transaction dataset
print("Original rows:", len(df))
print("Cleaned transaction rows:", len(df_clean))
print("Exact duplicates remaining:",
      df_clean.duplicated().sum())

print("Negative quantities retained:",
      (df_clean["Quantity"] < 0).sum())

print("Missing customer IDs retained:",
      df_clean["CustomerID"].isna().sum())

print("Missing dates remaining:",
      df_clean["InvoiceDate"].isna().sum())

Original rows: 541909
Cleaned transaction rows: 536641
Exact duplicates remaining: 0
Negative quantities retained: 10587
Missing customer IDs retained: 135037
Missing dates remaining: 0


In [35]:
print("Positive sales rows:", len(sales_clean))

print(
    "Invalid quantities:",
    (sales_clean["Quantity"] <= 0).sum()
)

print(
    "Invalid prices:",
    (sales_clean["UnitPrice"] <= 0).sum()
)

print(
    "Cancelled invoices:",
    sales_clean["is_cancelled"].sum()
)

Positive sales rows: 524878
Invalid quantities: 0
Invalid prices: 0
Cancelled invoices: 0


In [37]:
assert not sales_clean.duplicated().any()
assert (sales_clean["Quantity"] > 0).all()
assert (sales_clean["UnitPrice"] > 0).all()
assert (~sales_clean["is_cancelled"]).all()
assert sales_clean["InvoiceDate"].notna().all()
assert sales_clean["Description"].notna().all()

print("All positive-sales validation checks passed!")

All positive-sales validation checks passed!


In [38]:
# Save cleaned transactions, including flags
df_clean.to_csv(
    "../data/cleaned_transactions.csv",
    index=False
)

# Save positive sales for sales analysis
sales_clean.to_csv(
    "../data/cleaned_sales.csv",
    index=False
)

print("Cleaned datasets saved successfully!")

Cleaned datasets saved successfully!


In [39]:
cleaning_summary = pd.DataFrame({
    "Metric": [
        "Original records",
        "Exact duplicates removed",
        "Cleaned transaction records",
        "Records excluded from positive sales"
    ],
    "Count": [
        len(df),
        duplicates_removed,
        len(df_clean),
        len(df_clean) - len(sales_clean)
    ]
})

display(cleaning_summary)

,Metric,Count
0,Original records,541909
1,Exact duplicates removed,5268
2,Cleaned transaction records,536641
3,Records excluded from positive sales,11763


### 6.3 Cleaning Results

The data-cleaning process was performed using Pandas based on the findings from the data-quality investigation.

The cleaning process included:

1. Removing exact duplicate records to reduce the risk of double counting.
2. Preserving missing customer IDs because they are not essential for transaction-level revenue calculations.
3. Retaining unusual quantities and prices in the cleaned transaction dataset while adding flags for further investigation.
4. Identifying cancelled transactions using invoice identifiers.
5. Creating a separate positive-sales dataset for sales performance analysis.
6. Validating the resulting datasets and saving them as CSV files.

### 6.4 Cleaning Summary

* Original records: 541909
* Exact duplicate records removed: 5268
* Cleaned transaction records: 536641
* Positive-sales records: 11763

### 6.5 Conclusion

The cleaning process produced a cleaned transaction dataset and a separate positive-sales dataset. The original dataset was preserved, and unusual transactions were retained or flagged where appropriate to avoid losing potentially useful business information.

The resulting datasets are ready for feature engineering and further exploratory data analysis.


## 7. Feature Engineering

Feature engineering is the process of creating new variables from existing data to support analysis.

In this step, we will:

* Calculate transaction revenue.
* Extract year, month, and day from transaction dates.
* Create a year-month feature for monthly sales analysis.
* Prepare separate datasets for sales and returns analysis.

These features will help us identify sales trends, measure revenue, and understand customer purchasing activity.


In [52]:
# Calculate revenue for all cleaned transactions
df_clean["Revenue"] = (
    df_clean["Quantity"] * df_clean["UnitPrice"]
)

# Calculate revenue for positive sales
sales_clean["Revenue"] = (
    sales_clean["Quantity"] * sales_clean["UnitPrice"]
)

# Display the results
display(
    sales_clean[
        ["InvoiceNo", "Description", "Quantity",
         "UnitPrice", "Revenue"]
    ].head(10)
)

,InvoiceNo,Description,Quantity,UnitPrice,Revenue
0,536365,WHITE HANGING HEART T-LIGHT HOLDER,6,2.55,15.30
1,536365,WHITE METAL LANTERN,6,3.39,20.34
2,536365,CREAM CUPID HEARTS COAT HANGER,8,2.75,22.00
3,536365,KNITTED UNION FLAG HOT WATER BOTTLE,6,3.39,20.34
4,536365,RED WOOLLY HOTTIE WHITE HEART.,6,3.39,20.34
5,536365,SET 7 BABUSHKA NESTING BOXES,2,7.65,15.30
6,536365,GLASS STAR FROSTED T-LIGHT HOLDER,6,4.25,25.50
7,536366,HAND WARMER UNION JACK,6,1.85,11.10
8,536366,HAND WARMER RED POLKA DOT,6,1.85,11.10
9,536367,ASSORTED COLOUR BIRD ORNAMENT,32,1.69,54.08


In [53]:
# Ensure InvoiceDate is in datetime format
df_clean["InvoiceDate"] = pd.to_datetime(
    df_clean["InvoiceDate"], errors="coerce"
)

sales_clean["InvoiceDate"] = pd.to_datetime(
    sales_clean["InvoiceDate"], errors="coerce"
)

# Extract date features
for data in [df_clean, sales_clean]:
    data["Year"] = data["InvoiceDate"].dt.year
    data["Month"] = data["InvoiceDate"].dt.month
    data["MonthName"] = data["InvoiceDate"].dt.month_name()
    data["Day"] = data["InvoiceDate"].dt.day
    data["DayOfWeek"] = data["InvoiceDate"].dt.day_name()
    data["YearMonth"] = (
        data["InvoiceDate"].dt.to_period("M")
    )

# Preview engineered features
display(
    sales_clean[
        ["InvoiceDate", "Year", "Month", "MonthName",
         "Day", "DayOfWeek", "YearMonth", "Revenue"]
    ].head(10)
)

,InvoiceDate,Year,Month,MonthName,Day,DayOfWeek,YearMonth,Revenue
0,2010-12-01 08:26:00,2010,12,December,1,Wednesday,2010-12,15.30
1,2010-12-01 08:26:00,2010,12,December,1,Wednesday,2010-12,20.34
2,2010-12-01 08:26:00,2010,12,December,1,Wednesday,2010-12,22.00
3,2010-12-01 08:26:00,2010,12,December,1,Wednesday,2010-12,20.34
4,2010-12-01 08:26:00,2010,12,December,1,Wednesday,2010-12,20.34
5,2010-12-01 08:26:00,2010,12,December,1,Wednesday,2010-12,15.30
6,2010-12-01 08:26:00,2010,12,December,1,Wednesday,2010-12,25.50
7,2010-12-01 08:28:00,2010,12,December,1,Wednesday,2010-12,11.10
8,2010-12-01 08:28:00,2010,12,December,1,Wednesday,2010-12,11.10
9,2010-12-01 08:34:00,2010,12,December,1,Wednesday,2010-12,54.08


In [54]:
# Check new columns
feature_columns = [
    "Revenue", "Year", "Month", "MonthName",
    "Day", "DayOfWeek", "YearMonth"
]

display(sales_clean[feature_columns].isnull().sum())

# Check revenue calculation
assert np.allclose(
    sales_clean["Revenue"],
    sales_clean["Quantity"] * sales_clean["UnitPrice"]
)

print("Feature engineering validation completed.")

Revenue      0
Year         0
Month        0
MonthName    0
Day          0
DayOfWeek    0
YearMonth    0
dtype: int64

Feature engineering validation completed.


In [55]:
sales_clean.to_csv(
    "../data/engineered_sales.csv",
    index=False
)

print("Engineered sales dataset saved successfully!")

Engineered sales dataset saved successfully!


### Feature Engineering Summary

In this step, we created:

* **Revenue:** the value of each transaction line.
* **Year:** the year of the transaction.
* **Month:** the numerical month of the transaction.
* **MonthName:** the name of the month.
* **Day:** the day of the month.
* **DayOfWeek:** the weekday of the transaction.
* **YearMonth:** a monthly time period for trend analysis.

These features will be used in the next step to analyze sales performance, product performance, and geographic trends.


## 8. Exploratory Data Analysis (EDA)

Exploratory Data Analysis (EDA) is used to summarize and investigate a dataset to discover patterns, trends, and relationships.

In this step, we will analyze:

1. Overall sales performance.
2. Product performance.
3. Monthly sales trends.
4. Sales performance by country.
5. Returns and negative-quantity transactions.

The purpose is to answer business questions using the cleaned and engineered data rather than relying on assumptions.


In [56]:
# Overall sales performance
total_revenue = sales_clean["Revenue"].sum()

total_units_sold = sales_clean["Quantity"].sum()

total_orders = sales_clean["InvoiceNo"].nunique()

average_order_value = (
    sales_clean.groupby("InvoiceNo")["Revenue"].sum().mean()
)

# Net recorded revenue includes negative-quantity transactions
net_recorded_revenue = df_clean["Revenue"].sum()

sales_metrics = pd.DataFrame({
    "Metric": [
        "Gross positive sales revenue",
        "Net recorded revenue",
        "Total units sold",
        "Total orders",
        "Average order value"
    ],
    "Value": [
        total_revenue,
        net_recorded_revenue,
        total_units_sold,
        total_orders,
        average_order_value
    ]
})

display(sales_metrics)

,Metric,Value
0,Gross positive sales revenue,"10,642,110.80"
1,Net recorded revenue,"9,726,006.95"
2,Total units sold,"5,572,420.00"
3,Total orders,"19,960.00"
4,Average order value,533.17


In [58]:
# Top 10 products by quantity sold
top_products_quantity = (
    sales_clean.groupby("Description", as_index=False)
    .agg(
        Quantity_Sold=("Quantity", "sum"),
        Revenue=("Revenue", "sum")
    )
    .sort_values("Quantity_Sold", ascending=False)
    .head(10)
)

display(top_products_quantity)

,Description,Quantity_Sold,Revenue
2387,"PAPER CRAFT , LITTLE BIRDIE",80995,"168,469.60"
2052,MEDIUM CERAMIC TOP STORAGE JAR,78033,"81,700.92"
3934,WORLD WAR 2 GLIDERS ASSTD DESIGNS,54951,"13,814.01"
1816,JUMBO BAG RED RETROSPOT,48371,"94,159.81"
3844,WHITE HANGING HEART T-LIGHT HOLDER,37872,"106,236.72"
2681,POPCORN HOLDER,36749,"34,288.67"
2337,PACK OF 72 RETROSPOT CAKE CASES,36396,"21,246.45"
227,ASSORTED COLOUR BIRD ORNAMENT,36362,"58,927.62"
2741,RABBIT NIGHT LIGHT,30739,"66,870.03"
2107,MINI PAINT SET VINTAGE,26633,"16,937.82"


In [59]:
# Top 10 products by revenue
top_products_revenue = (
    sales_clean.groupby("Description", as_index=False)
    .agg(
        Revenue=("Revenue", "sum"),
        Quantity_Sold=("Quantity", "sum")
    )
    .sort_values("Revenue", ascending=False)
    .head(10)
)

display(top_products_revenue)

,Description,Revenue,Quantity_Sold
1067,DOTCOM POSTAGE,"206,248.77",706
2853,REGENCY CAKESTAND 3 TIER,"174,156.54",13851
2387,"PAPER CRAFT , LITTLE BIRDIE","168,469.60",80995
3844,WHITE HANGING HEART T-LIGHT HOLDER,"106,236.72",37872
2413,PARTY BUNTING,"99,445.23",18283
1816,JUMBO BAG RED RETROSPOT,"94,159.81",48371
2052,MEDIUM CERAMIC TOP STORAGE JAR,"81,700.92",78033
2692,POSTAGE,"78,101.88",3150
2192,Manual,"77,752.82",6985
2741,RABBIT NIGHT LIGHT,"66,870.03",30739


In [60]:
# Group revenue and orders by month
monthly_sales = (
    sales_clean.groupby("YearMonth", as_index=False)
    .agg(
        Revenue=("Revenue", "sum"),
        Orders=("InvoiceNo", "nunique"),
        Units_Sold=("Quantity", "sum")
    )
    .sort_values("YearMonth")
)

display(monthly_sales)

,YearMonth,Revenue,Orders,Units_Sold
0,2010-12,"821,452.73",1559,358019
1,2011-01,"689,811.61",1086,387099
2,2011-02,"522,545.56",1100,282934
3,2011-03,"716,215.26",1454,376599
4,2011-04,"536,968.49",1246,307953
5,2011-05,"769,296.61",1681,395001
6,2011-06,"760,547.01",1533,388511
7,2011-07,"718,076.12",1475,399693
8,2011-08,"757,841.38",1361,421020
9,2011-09,"1,056,435.19",1837,569573


In [61]:
# Month with the highest revenue
highest_sales_month = monthly_sales.loc[
    monthly_sales["Revenue"].idxmax()
]

print("Month with highest revenue:")
display(highest_sales_month)

Month with highest revenue:


YearMonth         2011-11
Revenue      1,503,866.78
Orders               2769
Units_Sold         751377
Name: 11, dtype: object

In [62]:
# Month with the lowest revenue
lowest_sales_month = monthly_sales.loc[
    monthly_sales["Revenue"].idxmin()
]

print("Month with lowest revenue:")
display(lowest_sales_month)

Month with lowest revenue:


YearMonth       2011-02
Revenue      522,545.56
Orders             1100
Units_Sold       282934
Name: 2, dtype: object

In [63]:
# Sales performance by country
country_sales = (
    sales_clean.groupby("Country", as_index=False)
    .agg(
        Revenue=("Revenue", "sum"),
        Orders=("InvoiceNo", "nunique"),
        Units_Sold=("Quantity", "sum")
    )
    .sort_values("Revenue", ascending=False)
)

# Display the top 10 countries
display(country_sales.head(10))

,Country,Revenue,Orders,Units_Sold
36,United Kingdom,"9,001,744.09",18019,4646906
24,Netherlands,"285,446.34",94,200361
10,EIRE,"283,140.52",288,147007
14,Germany,"228,678.40",457,119154
13,France,"209,625.37",392,112060
0,Australia,"138,453.81",57,83891
31,Spain,"61,558.56",90,27933
33,Switzerland,"57,067.60",54,30617
3,Belgium,"41,196.34",98,23237
32,Sweden,"38,367.83",36,36078


In [64]:
country_sales["Revenue_Percentage"] = (
    country_sales["Revenue"] /
    country_sales["Revenue"].sum() * 100
)

display(country_sales.head(10))

,Country,Revenue,Orders,Units_Sold,Revenue_Percentage
36,United Kingdom,"9,001,744.09",18019,4646906,84.59
24,Netherlands,"285,446.34",94,200361,2.68
10,EIRE,"283,140.52",288,147007,2.66
14,Germany,"228,678.40",457,119154,2.15
13,France,"209,625.37",392,112060,1.97
0,Australia,"138,453.81",57,83891,1.30
31,Spain,"61,558.56",90,27933,0.58
33,Switzerland,"57,067.60",54,30617,0.54
3,Belgium,"41,196.34",98,23237,0.39
32,Sweden,"38,367.83",36,36078,0.36


In [65]:
# Identify negative-quantity transactions
returns_df = df_clean[
    df_clean["Quantity"] < 0
].copy()

# Summarize negative-quantity transactions
returns_summary = pd.DataFrame({
    "Metric": [
        "Negative-quantity records",
        "Invoices containing negative quantities",
        "Total negative quantity",
        "Recorded value of negative-quantity records"
    ],
    "Value": [
        len(returns_df),
        returns_df["InvoiceNo"].nunique(),
        returns_df["Quantity"].sum(),
        returns_df["Revenue"].sum()
    ]
})

display(returns_summary)

,Metric,Value
0,Negative-quantity records,"10,587.00"
1,Invoices containing negative quantities,"5,172.00"
2,Total negative quantity,"-482,517.00"
3,Recorded value of negative-quantity records,"-893,979.73"


In [67]:
# Products with the largest negative recorded values
returns_by_product = (
    returns_df.groupby("Description", as_index=False)
    .agg(
        Negative_Quantity=("Quantity", "sum"),
        Recorded_Value=("Revenue", "sum")
    )
    .sort_values("Recorded_Value")
    .head(10)
)

display(returns_by_product)

,Description,Negative_Quantity,Recorded_Value
94,AMAZON FEE,-32,"-235,281.59"
1220,"PAPER CRAFT , LITTLE BIRDIE",-80995,"-168,469.60"
1126,Manual,-4066,"-146,784.46"
1056,MEDIUM CERAMIC TOP STORAGE JAR,-74494,"-77,479.64"
1342,POSTAGE,-147,"-11,871.24"
1437,REGENCY CAKESTAND 3 TIER,-855,"-9,697.05"
459,CRUK Commission,-16,"-7,933.43"
287,Bank Charges,-25,"-7,340.64"
1932,WHITE HANGING HEART T-LIGHT HOLDER,-2578,"-6,624.30"
621,FAIRY CAKE FLANNEL ASSORTED COLOUR,-3150,"-6,591.42"


### EDA Findings

**1. Overall Sales Performance**

* Total positive sales revenue: 10,642,110.80
* Total orders: 19,960
* Total units sold: 5,572,420
* Average order value: 533.17

**2. Product Performance**

* Highest-selling product by quantity: PAPER CRAFT, LITTLE BIRDIE
* Highest-revenue product: DOTCOM POSTAGE
* **Key observation:** The product with the highest sales quantity is different from the one generating the highest revenue. PAPER CRAFT, LITTLE BIRDIE leads in sales volume, while DOTCOM POSTAGE generates the most revenue. This demonstrates that sales volume and revenue do not necessarily follow the same pattern. The high revenue from DOTCOM POSTAGE may reflect its pricing or the number of transactions involving it.

**3. Monthly Sales**

* Highest-revenue month: 2011-11
* Lowest-revenue month: 2011-02
* **Key observation:** November 2011 recorded the highest revenue, while February 2011 recorded the lowest. This indicates that sales revenue varied across the observed months. The difference may be associated with changes in purchasing activity or seasonal demand, but further analysis is needed to establish the reasons for this variation.

**4. Country Performance**

* Highest-revenue country: United Kingdom
* Country with the most orders: United Kingdom
* **Key observation:** The United Kingdom leads in both revenue and order count, indicating that it is the main market in this dataset. This suggests that the business's recorded sales are concentrated in the UK. Further analysis of revenue percentages and other countries is needed to understand the extent of this concentration.

**5. Returns Analysis**

* Negative-quantity records: 10,587
* Total recorded value of negative-quantity transactions: -893,979.73
* **Key observation:** The dataset contains 10,587 negative-quantity records with a combined recorded value of -893,979.73. These transactions reduce the net recorded revenue compared with positive sales revenue. The negative values may represent returns or other transaction adjustments; further investigation is required to determine their exact causes.

### EDA Conclusion

The exploratory data analysis reveals several important patterns in the Online Retail dataset. The business recorded positive sales revenue of 10,642,110.80 across 19,960 orders, with an average order value of 533.17.

Product analysis shows that PAPER CRAFT, LITTLE BIRDIE had the highest sales quantity, whereas DOTCOM POSTAGE generated the highest revenue. This difference highlights the importance of examining both sales volume and revenue when evaluating product performance.

Monthly analysis indicates that November 2011 had the highest revenue, while February 2011 had the lowest. This shows variation in sales performance over the observed period, although additional analysis is required to establish the causes of these changes.

Geographic analysis shows that the United Kingdom had both the highest revenue and the largest number of orders, making it the leading country in the dataset. Meanwhile, returns analysis identified 10,587 negative-quantity records with a combined recorded value of -893,979.73, which should be considered when evaluating net recorded revenue.

Overall, the analysis identifies differences in product performance, variation in monthly sales, concentration of sales in the United Kingdom, and the presence of negative-quantity transactions. These findings provide a foundation for further investigation through data visualization and more detailed analysis. However, the results describe the observed transactions and do not, on their own, establish the causes of the patterns.
